### PR TITLE
CPKAFKA-1054 : introduce batch expiry config

### PR DESCRIPTION
**Pull request by apurvam**
We solve the problem of expired batches not being retried by providing the a separate config `batch.expiry.ms` which controls how long to wait before batches are expired. Before this patch, we would expire batches once `request.timeout.ms` elapsed. By setting the new config to be very high, we ensure that batches are never expired, which has the same effect as being retried indefinitely.

### CODE REVIEW

**Comment by apurvam** [comment]
cc @ijuma

**Comment by apurvam** [comment]
I am not sure what tests make sense to add since we have the exactly the same behavior as before, except that the batch expiry could be potentially different from the request timeout.

**Comment by ijuma** [comment]
With regards to tests, it seems like a bounce test of some sort is the only way we could truly test the behaviour difference. Maybe we can just add a unit tests that verifies that we configure the batch expiry timeout correctly when set and unset.

File: clients/src/test/java/org/apache/kafka/clients/producer/KafkaProducerTest.java [hunk 1]
```diff
@@ -484,4 +484,18 @@ public void testPartitionsForWithNullTopic() {
...
+        Properties props = new Properties();
+        props.setProperty(ProducerConfig.BOOTSTRAP_SERVERS_CONFIG, "localhost:2002");
+        props.setProperty(ProducerConfig.REQUEST_TIMEOUT_MS_CONFIG, "5");
+        KafkaProducer<String, String> producerWithDefaultBatchExpiry = new KafkaProducer<>(props, new StringSerializer(), new StringSerializer());
+
+        assertEquals(5, producerWithDefaultBatchExpiry.batchExpiryMs());
+        props.setProperty(ProducerConfig.BATCH_EXPIRY_MS, "10");
+
+        KafkaProducer<String, String> producerWithConfiguredBatchExpiry = new KafkaProducer<>(props, new StringSerializer(), new StringSerializer());
+        assertEquals(10L, producerWithConfiguredBatchExpiry.batchExpiryMs());
```
Comment (ijuma) [inline clients/src/test/java/org/apache/kafka/clients/producer/KafkaProducerTest.java]: We should close the producers after usage (maybe use try with resources).

Comment (apurvam) [reply clients/src/test/java/org/apache/kafka/clients/producer/KafkaProducerTest.java]: ack. Updated the patch.

**Comment by ijuma** [comment]
Should we add a `confluent.` prefix to the config name to make it clear that it's not to be used?

**Comment by ijuma** [comment]
@apurvam, what do you think of my suggestion above?

**Comment by apurvam** [comment]
Sorry, I missed that message @ijuma .. Sure, we can add the 'confluent' prefix, though that doesn't necessarily indicate that it isn't to be used. In all honesty, this setting can very well be used without much ill effects. 

But I agree that adding the confluent prefix is a good way to call out that this is something that is not standard. I will add it.

**Comment by ijuma** [comment]
Yes, I just wanted to make it clear that it's non-standard as we don't intend to maintain it for longer than necessary.

**Comment by apurvam** [comment]
I guess the only remaining question is how to merge this? Should I create a patch file or merge it directly onto the 0.11.0 branch of confluentinc/kafka?

**Pull request merged**